edb/edgeql/compiler/setgen.py

```diff
@@ -1346,6 +1351,51 @@ def computable_ptr_set(
     return comp_ir_set
 
 
+def _get_schema_computed_ctx(
+    *,
+    rptr: irast.Pointer,
+    source: irast.Set,
+    ctx: context.ContextLevel
+) -> Callable[[], ContextManager[context.ContextLevel]]:
+
+    @contextlib.contextmanager
+    def newctx() -> Iterator[context.ContextLevel]:
+        with ctx.new() as subctx:
+            subctx.class_view_overrides = {}
+            subctx.partial_path_prefix = None
+
+            source_stype = get_set_type(source, ctx=ctx)
+            subctx.view_map = ctx.view_map.new_child()
+
+            source_scope = pathctx.get_set_scope(rptr.source, ctx=ctx)
+            if source_scope and source_scope.namespaces:
+                subctx.path_id_namespace |= source_scope.namespaces
+
+            pending_pid_ns = {ctx.aliases.get('ns')}
+
+            subctx.pending_stmt_own_path_id_namespace = (
+                frozenset(pending_pid_ns))
+
+            subns = set(pending_pid_ns)
+            subns.update(source.path_id.namespace)
+
+            inner_path_id = pathctx.get_path_id(
+                source_stype, ctx=subctx).merge_namespace(subns)
+
+            subctx.pending_stmt_full_path_id_namespace = frozenset(subns)
+
+            remapped_source = new_set_from_set(
+                rptr.source, rptr=rptr.source.rptr,
+                preserve_scope_ns=True, ctx=ctx)
+            key = inner_path_id.strip_namespace(inner_path_id.namespace)
+            old = subctx.view_map.get(key, ())
+            subctx.view_map[key] = ((inner_path_id, remapped_source),) + old
```

Comment:
You still want this to be `detached` and you only need to populate the `view_map` in the detached context, so something like this:

```suggestion
        with ctx.detached() as subctx:
            source_stype = get_set_type(source, ctx=ctx)
            subctx.view_map = ctx.view_map.new_child()

            source_scope = pathctx.get_set_scope(rptr.source, ctx=ctx)
            if source_scope and source_scope.namespaces:
                subctx.path_id_namespace |= source_scope.namespaces

            inner_path_id = pathctx.get_path_id(source_stype, ctx=subctx)

            remapped_source = new_set_from_set(
                rptr.source,
                rptr=rptr.source.rptr,
                preserve_scope_ns=True,
                ctx=ctx,
            )
            key = inner_path_id.strip_namespace(inner_path_id.namespace)
            old = subctx.view_map.get(key, ())
            subctx.view_map[key] = ((inner_path_id, remapped_source),) + old
```

Reply:
Updated.

Reply:
Actually, even this version is too liberal, you don't want to preserve previous values of `view_map` in the detached context either, it should be exactly one entry.  Apply this patch:

```patch
diff --git a/edb/edgeql/compiler/context.py b/edb/edgeql/compiler/context.py
index 956806aa7..7252f82c6 100644
--- a/edb/edgeql/compiler/context.py
+++ b/edb/edgeql/compiler/context.py
@@ -658,6 +658,7 @@ class ContextLevel(compiler.ContextLevel):
                 self.anchors = prevlevel.anchors.copy()
                 self.modaliases = prevlevel.modaliases.copy()
                 self.aliased_views = collections.ChainMap()
+                self.view_map = collections.ChainMap()
                 self.class_view_overrides = {}
                 self.expr_exposed = prevlevel.expr_exposed

diff --git a/edb/edgeql/compiler/setgen.py b/edb/edgeql/compiler/setgen.py
index 530df46b9..66ee0bb2c 100644
--- a/edb/edgeql/compiler/setgen.py
+++ b/edb/edgeql/compiler/setgen.py
@@ -1357,8 +1357,6 @@ def _get_schema_computed_ctx(
     def newctx() -> Iterator[context.ContextLevel]:
         with ctx.detached() as subctx:
             source_stype = get_set_type(source, ctx=ctx)
-            subctx.view_map = ctx.view_map.new_child()
-
             source_scope = pathctx.get_set_scope(rptr.source, ctx=ctx)
             if source_scope and source_scope.namespaces:
                 subctx.path_id_namespace |= source_scope.namespaces
@@ -1372,8 +1370,7 @@ def _get_schema_computed_ctx(
                 ctx=ctx,
             )
             key = inner_path_id.strip_namespace(inner_path_id.namespace)
-            old = subctx.view_map.get(key, ())
-            subctx.view_map[key] = ((inner_path_id, remapped_source),) + old
+            subctx.view_map[key] = ((inner_path_id, remapped_source),)

             yield subctx

```

May also be worth writing a query that has a computed link referring to the type of an enclosing shape, e.g.:

```
SELECT Foo {
   bar: {
      bar_comp
   }
}
```

where `bar_comp` has references to `Foo`.  These *must not* be correlated with `Foo` in the query.

Reply:
> May also be worth writing a query that has a computed link referring to the type of an enclosing shape

Oh, cool, I'll add a test for that.

Reply:
> Actually, even this version is too liberal, you don't want to preserve previous values of `view_map` in the detached context either, it should be exactly one entry.

So... this "previous value" of `view_map` would be coming from where exactly? A `WITH` block of the `CREATE TYPE` command? I'm trying to understand a way to _have_ some previous values there that would not be valid in the nested scope.

Reply:
Nested computables, e.g. if you turn the above test to:

```
SELECT Spam {
   foo := (SELECT Foo {
     bar := (SELECT Bar {
        bar_comp
     })
   })
}
```

`bar_comp` should only have a mapping for `Bar -> Spam.foo.bar`, but not for `Foo -> Spam.foo`